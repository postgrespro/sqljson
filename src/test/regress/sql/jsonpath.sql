--jsonpath io

select '$'::jsonpath;
select 'strict $'::jsonpath;
select 'lax $'::jsonpath;
select '$.a'::jsonpath;
select '$.a.v'::jsonpath;
select '$.a.*'::jsonpath;
select '$.*.[*]'::jsonpath;
select '$.*[*]'::jsonpath;
select '$.a.[*]'::jsonpath;
select '$.a[*]'::jsonpath;
select '$.a.[*][*]'::jsonpath;
select '$.a.[*].[*]'::jsonpath;
select '$.a[*][*]'::jsonpath;
select '$.a[*].[*]'::jsonpath;
select '$[*]'::jsonpath;
select '$[0]'::jsonpath;
select '$[*][0]'::jsonpath;
select '$[*].a'::jsonpath;
select '$[*][0].a.b'::jsonpath;
select '$.a.**.b'::jsonpath;
select '$.a.**{2}.b'::jsonpath;
select '$.a.**{2,2}.b'::jsonpath;
select '$.a.**{2,5}.b'::jsonpath;
select '$.a.**{,5}.b'::jsonpath;
select '$.a.**{5,}.b'::jsonpath;
select '$+1'::jsonpath;
select '$-1'::jsonpath;
select '$--+1'::jsonpath;
select '$.a/+-1'::jsonpath;

select '$.g ? (@ == 1)'::jsonpath;
select '$.g ? (a == 1)'::jsonpath;
select '$.g ? (.a == 1)'::jsonpath;
select '$.g ? (@.a == 1)'::jsonpath;
select '$.g ? (@.a == 1 || a == 4)'::jsonpath;
select '$.g ? (@.a == 1 && a == 4)'::jsonpath;
select '$.g ? (@.a == 1 || a == 4 && b == 7)'::jsonpath;
select '$.g ? (@.a == 1 || !(a == 4) && b == 7)'::jsonpath;
select '$.g ? (@.a == 1 || !(x >= 123 || a == 4) && b == 7)'::jsonpath;
select '$.g ? (.x >= @[*]?(@.a > "abc"))'::jsonpath;
select '$.g ? ((x >= 123 || a == 4) is unknown)'::jsonpath;
select '$.g ? (exists (.x))'::jsonpath;
select '$.g ? (exists (@.x ? (@ == 14)))'::jsonpath;
select '$.g ? (exists (.x ? (@ == 14)))'::jsonpath;
select '$.g ? ((x >= 123 || a == 4) && exists (.x ? (@ == 14)))'::jsonpath;
select '$.g ? (+x >= +-(+a + 2))'::jsonpath;

select '$a'::jsonpath;
select '$a.b'::jsonpath;
select '$a[*]'::jsonpath;
select '$.g ? (zip == $zip)'::jsonpath;
select '$.a.[1,2, 3 to 16]'::jsonpath;
select '$.a[1,2, 3 to 16]'::jsonpath;
select '$.a[$a + 1, ($b[*]) to -(@[0] * 2)]'::jsonpath;
select 'last'::jsonpath;
select '"last"'::jsonpath;
select '$.last'::jsonpath;
select '$ ? (last > 0)'::jsonpath;
select '$[last]'::jsonpath;
select '$[@ ? (last > 0)]'::jsonpath;

select 'null.type()'::jsonpath;
select '1.type()'::jsonpath;
select '"aaa".type()'::jsonpath;
select 'aaa.type()'::jsonpath;
select 'true.type()'::jsonpath;
select '$.datetime()'::jsonpath;
select '$.datetime("datetime template")'::jsonpath;

select '$ ? (@ starts with "abc")'::jsonpath;
select '$ ? (@ starts with $var)'::jsonpath;

select '$ ? (@ like_regex "pattern")'::jsonpath;
select '$ ? (@ like_regex "pattern" flag "")'::jsonpath;
select '$ ? (@ like_regex "pattern" flag "i")'::jsonpath;
select '$ ? (@ like_regex "pattern" flag "is")'::jsonpath;
select '$ ? (@ like_regex "pattern" flag "isim")'::jsonpath;
select '$ ? (@ like_regex "pattern" flag "xsms")'::jsonpath;
select '$ ? (@ like_regex "pattern" flag "a")'::jsonpath;

select '$ ? (a < 1)'::jsonpath;
select '$ ? (a < -1)'::jsonpath;
select '$ ? (a < +1)'::jsonpath;
select '$ ? (a < .1)'::jsonpath;
select '$ ? (a < -.1)'::jsonpath;
select '$ ? (a < +.1)'::jsonpath;
select '$ ? (a < 0.1)'::jsonpath;
select '$ ? (a < -0.1)'::jsonpath;
select '$ ? (a < +0.1)'::jsonpath;
select '$ ? (a < 10.1)'::jsonpath;
select '$ ? (a < -10.1)'::jsonpath;
select '$ ? (a < +10.1)'::jsonpath;
select '$ ? (a < 1e1)'::jsonpath;
select '$ ? (a < -1e1)'::jsonpath;
select '$ ? (a < +1e1)'::jsonpath;
select '$ ? (a < .1e1)'::jsonpath;
select '$ ? (a < -.1e1)'::jsonpath;
select '$ ? (a < +.1e1)'::jsonpath;
select '$ ? (a < 0.1e1)'::jsonpath;
select '$ ? (a < -0.1e1)'::jsonpath;
select '$ ? (a < +0.1e1)'::jsonpath;
select '$ ? (a < 10.1e1)'::jsonpath;
select '$ ? (a < -10.1e1)'::jsonpath;
select '$ ? (a < +10.1e1)'::jsonpath;
select '$ ? (a < 1e-1)'::jsonpath;
select '$ ? (a < -1e-1)'::jsonpath;
select '$ ? (a < +1e-1)'::jsonpath;
select '$ ? (a < .1e-1)'::jsonpath;
select '$ ? (a < -.1e-1)'::jsonpath;
select '$ ? (a < +.1e-1)'::jsonpath;
select '$ ? (a < 0.1e-1)'::jsonpath;
select '$ ? (a < -0.1e-1)'::jsonpath;
select '$ ? (a < +0.1e-1)'::jsonpath;
select '$ ? (a < 10.1e-1)'::jsonpath;
select '$ ? (a < -10.1e-1)'::jsonpath;
select '$ ? (a < +10.1e-1)'::jsonpath;
select '$ ? (a < 1e+1)'::jsonpath;
select '$ ? (a < -1e+1)'::jsonpath;
select '$ ? (a < +1e+1)'::jsonpath;
select '$ ? (a < .1e+1)'::jsonpath;
select '$ ? (a < -.1e+1)'::jsonpath;
select '$ ? (a < +.1e+1)'::jsonpath;
select '$ ? (a < 0.1e+1)'::jsonpath;
select '$ ? (a < -0.1e+1)'::jsonpath;
select '$ ? (a < +0.1e+1)'::jsonpath;
select '$ ? (a < 10.1e+1)'::jsonpath;
select '$ ? (a < -10.1e+1)'::jsonpath;
select '$ ? (a < +10.1e+1)'::jsonpath;
