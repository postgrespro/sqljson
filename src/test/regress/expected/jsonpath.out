--jsonpath io
select '$'::jsonpath;
 jsonpath 
----------
 $
(1 row)

select 'strict $'::jsonpath;
 jsonpath 
----------
 strict $
(1 row)

select 'lax $'::jsonpath;
 jsonpath 
----------
 $
(1 row)

select '$.a'::jsonpath;
 jsonpath 
----------
 $."a"
(1 row)

select '$.a.v'::jsonpath;
 jsonpath  
-----------
 $."a"."v"
(1 row)

select '$.a.*'::jsonpath;
 jsonpath 
----------
 $."a".*
(1 row)

select '$.*.[*]'::jsonpath;
 jsonpath 
----------
 $.*[*]
(1 row)

select '$.*[*]'::jsonpath;
 jsonpath 
----------
 $.*[*]
(1 row)

select '$.a.[*]'::jsonpath;
 jsonpath 
----------
 $."a"[*]
(1 row)

select '$.a[*]'::jsonpath;
 jsonpath 
----------
 $."a"[*]
(1 row)

select '$.a.[*][*]'::jsonpath;
  jsonpath   
-------------
 $."a"[*][*]
(1 row)

select '$.a.[*].[*]'::jsonpath;
  jsonpath   
-------------
 $."a"[*][*]
(1 row)

select '$.a[*][*]'::jsonpath;
  jsonpath   
-------------
 $."a"[*][*]
(1 row)

select '$.a[*].[*]'::jsonpath;
  jsonpath   
-------------
 $."a"[*][*]
(1 row)

select '$[*]'::jsonpath;
 jsonpath 
----------
 $[*]
(1 row)

select '$[0]'::jsonpath;
 jsonpath 
----------
 $[0]
(1 row)

select '$[*][0]'::jsonpath;
 jsonpath 
----------
 $[*][0]
(1 row)

select '$[*].a'::jsonpath;
 jsonpath 
----------
 $[*]."a"
(1 row)

select '$[*][0].a.b'::jsonpath;
    jsonpath     
-----------------
 $[*][0]."a"."b"
(1 row)

select '$.a.**.b'::jsonpath;
   jsonpath   
--------------
 $."a".**."b"
(1 row)

select '$.a.**{2}.b'::jsonpath;
    jsonpath     
-----------------
 $."a".**{2}."b"
(1 row)

select '$.a.**{2,2}.b'::jsonpath;
    jsonpath     
-----------------
 $."a".**{2}."b"
(1 row)

select '$.a.**{2,5}.b'::jsonpath;
     jsonpath      
-------------------
 $."a".**{2,5}."b"
(1 row)

select '$.a.**{,5}.b'::jsonpath;
     jsonpath     
------------------
 $."a".**{,5}."b"
(1 row)

select '$.a.**{5,}.b'::jsonpath;
     jsonpath     
------------------
 $."a".**{5,}."b"
(1 row)

select '$+1'::jsonpath;
 jsonpath 
----------
 ($ + 1)
(1 row)

select '$-1'::jsonpath;
 jsonpath 
----------
 ($ - 1)
(1 row)

select '$--+1'::jsonpath;
 jsonpath 
----------
 ($ - -1)
(1 row)

select '$.a/+-1'::jsonpath;
   jsonpath   
--------------
 ($."a" / -1)
(1 row)

select '$.g ? (@ == 1)'::jsonpath;
    jsonpath    
----------------
 $."g"?(@ == 1)
(1 row)

select '$.g ? (a == 1)'::jsonpath;
     jsonpath     
------------------
 $."g"?("a" == 1)
(1 row)

select '$.g ? (.a == 1)'::jsonpath;
      jsonpath      
--------------------
 $."g"?(@."a" == 1)
(1 row)

select '$.g ? (@.a == 1)'::jsonpath;
      jsonpath      
--------------------
 $."g"?(@."a" == 1)
(1 row)

select '$.g ? (@.a == 1 || a == 4)'::jsonpath;
            jsonpath            
--------------------------------
 $."g"?(@."a" == 1 || "a" == 4)
(1 row)

select '$.g ? (@.a == 1 && a == 4)'::jsonpath;
            jsonpath            
--------------------------------
 $."g"?(@."a" == 1 && "a" == 4)
(1 row)

select '$.g ? (@.a == 1 || a == 4 && b == 7)'::jsonpath;
                  jsonpath                  
--------------------------------------------
 $."g"?(@."a" == 1 || "a" == 4 && "b" == 7)
(1 row)

select '$.g ? (@.a == 1 || !(a == 4) && b == 7)'::jsonpath;
                   jsonpath                    
-----------------------------------------------
 $."g"?(@."a" == 1 || !("a" == 4) && "b" == 7)
(1 row)

select '$.g ? (@.a == 1 || !(x >= 123 || a == 4) && b == 7)'::jsonpath;
                          jsonpath                           
-------------------------------------------------------------
 $."g"?(@."a" == 1 || !("x" >= 123 || "a" == 4) && "b" == 7)
(1 row)

select '$.g ? (.x >= @[*]?(@.a > "abc"))'::jsonpath;
               jsonpath                
---------------------------------------
 $."g"?(@."x" >= @[*]?(@."a" > "abc"))
(1 row)

select '$.g ? ((x >= 123 || a == 4) is unknown)'::jsonpath;
                  jsonpath                   
---------------------------------------------
 $."g"?(("x" >= 123 || "a" == 4) is unknown)
(1 row)

select '$.g ? (exists (.x))'::jsonpath;
        jsonpath        
------------------------
 $."g"?(exists (@."x"))
(1 row)

select '$.g ? (exists (@.x ? (@ == 14)))'::jsonpath;
             jsonpath             
----------------------------------
 $."g"?(exists (@."x"?(@ == 14)))
(1 row)

select '$.g ? (exists (.x ? (@ == 14)))'::jsonpath;
             jsonpath             
----------------------------------
 $."g"?(exists (@."x"?(@ == 14)))
(1 row)

select '$.g ? ((x >= 123 || a == 4) && exists (.x ? (@ == 14)))'::jsonpath;
                           jsonpath                           
--------------------------------------------------------------
 $."g"?(("x" >= 123 || "a" == 4) && exists (@."x"?(@ == 14)))
(1 row)

select '$.g ? (+x >= +-(+a + 2))'::jsonpath;
            jsonpath            
--------------------------------
 $."g"?(+"x" >= +(-(+"a" + 2)))
(1 row)

select '$a'::jsonpath;
 jsonpath 
----------
 $"a"
(1 row)

select '$a.b'::jsonpath;
 jsonpath 
----------
 $"a"."b"
(1 row)

select '$a[*]'::jsonpath;
 jsonpath 
----------
 $"a"[*]
(1 row)

select '$.g ? (zip == $zip)'::jsonpath;
        jsonpath         
-------------------------
 $."g"?("zip" == $"zip")
(1 row)

select '$.a.[1,2, 3 to 16]'::jsonpath;
      jsonpath      
--------------------
 $."a"[1,2,3 to 16]
(1 row)

select '$.a[1,2, 3 to 16]'::jsonpath;
      jsonpath      
--------------------
 $."a"[1,2,3 to 16]
(1 row)

select '$.a[$a + 1, ($b[*]) to -(@[0] * 2)]'::jsonpath;
                jsonpath                
----------------------------------------
 $."a"[$"a" + 1,$"b"[*] to -(@[0] * 2)]
(1 row)

select 'last'::jsonpath;
ERROR:  LAST is allowed only in array subscripts
LINE 1: select 'last'::jsonpath;
               ^
select '"last"'::jsonpath;
 jsonpath 
----------
 "last"
(1 row)

select '$.last'::jsonpath;
 jsonpath 
----------
 $."last"
(1 row)

select '$ ? (last > 0)'::jsonpath;
ERROR:  LAST is allowed only in array subscripts
LINE 1: select '$ ? (last > 0)'::jsonpath;
               ^
select '$[last]'::jsonpath;
 jsonpath 
----------
 $[last]
(1 row)

select '$[@ ? (last > 0)]'::jsonpath;
    jsonpath     
-----------------
 $[@?(last > 0)]
(1 row)

select 'null.type()'::jsonpath;
  jsonpath   
-------------
 null.type()
(1 row)

select '1.type()'::jsonpath;
 jsonpath 
----------
 1.type()
(1 row)

select '"aaa".type()'::jsonpath;
   jsonpath   
--------------
 "aaa".type()
(1 row)

select 'aaa.type()'::jsonpath;
   jsonpath   
--------------
 "aaa".type()
(1 row)

select 'true.type()'::jsonpath;
  jsonpath   
-------------
 true.type()
(1 row)

select '$.datetime()'::jsonpath;
   jsonpath   
--------------
 $.datetime()
(1 row)

select '$.datetime("datetime template")'::jsonpath;
            jsonpath             
---------------------------------
 $.datetime("datetime template")
(1 row)

select '$ ? (@ starts with "abc")'::jsonpath;
        jsonpath         
-------------------------
 $?(@ starts with "abc")
(1 row)

select '$ ? (@ starts with $var)'::jsonpath;
         jsonpath         
--------------------------
 $?(@ starts with $"var")
(1 row)

select '$ ? (@ like_regex "pattern")'::jsonpath;
          jsonpath          
----------------------------
 $?(@ like_regex "pattern")
(1 row)

select '$ ? (@ like_regex "pattern" flag "")'::jsonpath;
          jsonpath          
----------------------------
 $?(@ like_regex "pattern")
(1 row)

select '$ ? (@ like_regex "pattern" flag "i")'::jsonpath;
              jsonpath               
-------------------------------------
 $?(@ like_regex "pattern" flag "i")
(1 row)

select '$ ? (@ like_regex "pattern" flag "is")'::jsonpath;
               jsonpath               
--------------------------------------
 $?(@ like_regex "pattern" flag "is")
(1 row)

select '$ ? (@ like_regex "pattern" flag "isim")'::jsonpath;
               jsonpath               
--------------------------------------
 $?(@ like_regex "pattern" flag "im")
(1 row)

select '$ ? (@ like_regex "pattern" flag "xsms")'::jsonpath;
               jsonpath               
--------------------------------------
 $?(@ like_regex "pattern" flag "sx")
(1 row)

select '$ ? (@ like_regex "pattern" flag "a")'::jsonpath;
ERROR:  bad jsonpath representation
LINE 1: select '$ ? (@ like_regex "pattern" flag "a")'::jsonpath;
               ^
DETAIL:  unrecognized flag of LIKE_REGEX predicate at or near """
select '$ ? (a < 1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < -1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < -1)
(1 row)

select '$ ? (a < +1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < .1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 0.1)
(1 row)

select '$ ? (a < -.1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < -0.1)
(1 row)

select '$ ? (a < +.1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 0.1)
(1 row)

select '$ ? (a < 0.1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 0.1)
(1 row)

select '$ ? (a < -0.1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < -0.1)
(1 row)

select '$ ? (a < +0.1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 0.1)
(1 row)

select '$ ? (a < 10.1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < 10.1)
(1 row)

select '$ ? (a < -10.1)'::jsonpath;
    jsonpath     
-----------------
 $?("a" < -10.1)
(1 row)

select '$ ? (a < +10.1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < 10.1)
(1 row)

select '$ ? (a < 1e1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < 10)
(1 row)

select '$ ? (a < -1e1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < -10)
(1 row)

select '$ ? (a < +1e1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < 10)
(1 row)

select '$ ? (a < .1e1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < -.1e1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < -1)
(1 row)

select '$ ? (a < +.1e1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < 0.1e1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < -0.1e1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < -1)
(1 row)

select '$ ? (a < +0.1e1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < 10.1e1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 101)
(1 row)

select '$ ? (a < -10.1e1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < -101)
(1 row)

select '$ ? (a < +10.1e1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 101)
(1 row)

select '$ ? (a < 1e-1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 0.1)
(1 row)

select '$ ? (a < -1e-1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < -0.1)
(1 row)

select '$ ? (a < +1e-1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 0.1)
(1 row)

select '$ ? (a < .1e-1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < 0.01)
(1 row)

select '$ ? (a < -.1e-1)'::jsonpath;
    jsonpath     
-----------------
 $?("a" < -0.01)
(1 row)

select '$ ? (a < +.1e-1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < 0.01)
(1 row)

select '$ ? (a < 0.1e-1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < 0.01)
(1 row)

select '$ ? (a < -0.1e-1)'::jsonpath;
    jsonpath     
-----------------
 $?("a" < -0.01)
(1 row)

select '$ ? (a < +0.1e-1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < 0.01)
(1 row)

select '$ ? (a < 10.1e-1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < 1.01)
(1 row)

select '$ ? (a < -10.1e-1)'::jsonpath;
    jsonpath     
-----------------
 $?("a" < -1.01)
(1 row)

select '$ ? (a < +10.1e-1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < 1.01)
(1 row)

select '$ ? (a < 1e+1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < 10)
(1 row)

select '$ ? (a < -1e+1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < -10)
(1 row)

select '$ ? (a < +1e+1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < 10)
(1 row)

select '$ ? (a < .1e+1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < -.1e+1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < -1)
(1 row)

select '$ ? (a < +.1e+1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < 0.1e+1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < -0.1e+1)'::jsonpath;
   jsonpath   
--------------
 $?("a" < -1)
(1 row)

select '$ ? (a < +0.1e+1)'::jsonpath;
  jsonpath   
-------------
 $?("a" < 1)
(1 row)

select '$ ? (a < 10.1e+1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 101)
(1 row)

select '$ ? (a < -10.1e+1)'::jsonpath;
    jsonpath    
----------------
 $?("a" < -101)
(1 row)

select '$ ? (a < +10.1e+1)'::jsonpath;
   jsonpath    
---------------
 $?("a" < 101)
(1 row)

